src/main/java/org/apache/hadoop/security/UserGroupInformation.java

```diff
@@ -0,0 +1,2283 @@
+        }
+    }
+
+    /**
+     * Return the hash of the subject.
+     */
+    @Override
+    public int hashCode() {
+        return System.identityHashCode(subject);
+    }
+
+    /**
+     * Get the underlying subject from this ugi.
+     * @return the subject that represents this user.
+     */
+    public Subject getSubject() {
+        return subject;
+    }
+
+    /**
+     * Run the given action as the user, potentially throwing an exception.
+     * @param <T> the return type of the run method
+     * @param action the method to execute
+     * @return the value from the run method
+     * @throws IOException if the action throws an IOException
+     * @throws Error if the action throws an Error
+     * @throws RuntimeException if the action throws a RuntimeException
+     * @throws InterruptedException if the action throws an InterruptedException
+     * @throws UndeclaredThrowableException if the action throws something else
+     */
+    @InterfaceAudience.Public
+    @InterfaceStability.Evolving
+    public <T> T callAs(Callable<T> action
+    ) throws IOException, InterruptedException {
+        try {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("PrivilegedAction [as: {}][action: {}]", this, action,
+                        new Exception());
+            }
+            return Subject.callAs(subject, action);
+        } catch (CompletionException coe) {
+            Throwable cause = coe.getCause();
+            LOG.debug("CompletionException as: {}", this, cause);
+            if (cause == null) {
+                throw new RuntimeException("CompletionException with no " +
+                        "underlying cause. UGI [" + this + "]" +": " + coe, coe);
+            } else if (cause instanceof IOException) {
+                throw (IOException) cause;
+            } else if (cause instanceof Error) {
+                throw (Error) cause;
+            } else if (cause instanceof RuntimeException) {
+                throw (RuntimeException) cause;
+            } else if (cause instanceof InterruptedException) {
+                throw (InterruptedException) cause;
+            } else {
+                throw new UndeclaredThrowableException(cause);
+            }
+        }
+    }
+
+    /**
+     * Run the given action as the user, potentially throwing an exception.
+     * @param <T> the return type of the run method
+     * @param action the method to execute
+     * @return the value from the run method
+     * @throws IOException if the action throws an IOException
+     * @throws Error if the action throws an Error
+     * @throws RuntimeException if the action throws a RuntimeException
+     * @throws InterruptedException if the action throws an InterruptedException
+     * @throws UndeclaredThrowableException if the action throws something else
+     */
+    @InterfaceAudience.Public
+    @InterfaceStability.Evolving
+    public <T> T doAs(PrivilegedExceptionAction<T> action
```

Comment:
@wendigo  - this class doesn't include existing public method `UserGroupInformation.doAs(PrivilegedAction<T> action)`, which would break any library using this method. Any specific reason to exclude this method? If not, can you please consider adding this method?

https://hadoop.apache.org/docs/stable/api/org/apache/hadoop/security/UserGroupInformation.html#doAs-java.security.PrivilegedAction-

Reply:
This Hadoop shading project is only intended for use by Trino, which doesn't use this method. Can you explain more about your use case?

Reply:
We removed the `PrivilegedExceptionAction` version of this method as it uses `javax.security.auth.Subject#doAs(javax.security.auth.Subject, java.security.PrivilegedExceptionAction<T>)` which is deprecated for removal as of Java 18.

Reply:
@mneethiraj this is an internal Trino fork, that is only used by Trino afaik. We are on JDK 22 and will switch to 23 soon, so this repo contains fixes and changes that are required by the minimum Java version we use.

Reply:
> This Hadoop shading project is only intended for use by Trino, which doesn't use this method. Can you explain more about your use case?

Yes. a library used in Ranger authorizer for Trino uses `UserGroupInformation.doAs(PrivilegedAction<T> action)` while calling Apache Ranger to get policies to authorize accesses in Trino. Move to io.trino.hadoop:apache-hadoop version 3.3.5-3 (from 3.3.5-1) broke the authorizer - due to the missing/remoed method `UserGroupInformation.doAs(PrivilegedAction<T> action)` in this version.

Reply:
> @mneethiraj this is an internal Trino fork, that is only used by Trino afaik. We are on JDK 22 and will switch to 23 soon, so this repo contains fixes and changes that are required by the minimum Java version we use.

@wendigo  - is it not possible to retain method `UserGroupInformation.doAs(PrivilegedAction<T> action)` while having its implementation updated to work with JDK 23?

Reply:
We do not plan to

Reply:
any compelling reason??

Reply:
These method are deprecated and will be removed so there is no reason to keep them in the fork